File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-9.js
class Solution {
    /**
     * @param {string} s
     * @return {number}
     */
    lengthOfLongestSubstring(s) {
        let longest = 0
        let l = 0, r = 0
        const seen = new Set()

        while (r < s.length) {
            while (seen.has(s[r])) {
                seen.delete(s[l])
                l++
            }

            longest = Math.max(longest, r - l + 1)
            seen.add(s[r])
            r++
        }
        return longest
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-9.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        maxf = 0
        longest = 0
        count = {}
        l, r = 0, 0

        while r < len(s):
            count[s[r]] = 1 + count.get(s[r], 0)
            maxf = max(maxf, count[s[r]])

            while (r - l + 1) - maxf > k:
                count[s[l]] -= 1
                l += 1
            
            longest = max(longest, r - l + 1)
            r += 1
        return longest